### ISSUE TITLE
DynamoDB item creation using a aws_dynamodb_table_item resource

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-aws#517]
_This issue was originally opened by @drieselliott as hashicorp/terraform#11773. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi,

It would be nice to have a 'aws_dynamodb_table_item' resource that allows you to write configuration entries to DynamoDB tables when provisioning an environment. It should support both the DynamoDB JSON and the normal JSON format.

```hcl
resource "aws_dynamodb_table_item" "configuration_item" {
    table = "${aws_dynamodb_table.configuration_table.name}"
    format = "json"
    item = <<EOF
       //Some JSON content
EOF
}
```

table: The table name
format: Either 'json' or 'dynamodb' for 'normal json' or 'dynamodb json'
item: The item to insert into the table (either as inline content or template)

This is very useful in serverless applications that use DynamoDb as their configuration store and Terraform to provision everything. Otherwise you need to rely on a separate non-terraform script to upload your configuration entries.

Thanks for considering this.

Dries

**Comment by galadriel2143** [comment]
Started a PR into implement this feature. Mostly works but there are a few questions I would have before it gets merged back in.

Behavior on update: Should it force attributes back to the ones specified in the resource definition of they don't match? I think this would be fine based on how terraform handles attributes on other resources.
Behavior on externally removed item: Recreate, leave, alone, or hard fail? Recreating when it's missing seems to fall in line with other resources.

Output values: Currently I'm returning the range attribute value, hash attribute value, and the "query key", which contains both and can be used in a DDB query to get the object if necessary. Is there anything else that should be returned, or maybe we shouldn't return these at all...

Query setup: I have the range_key and hash_key names specified on the resource. These probably could be determined by querying the table metadata, but that would add more HTTP calls, and specifying them on the resource theoretically offers some protection from the key schema changing in an unexpected way.

**Comment by jantman** [comment]
Is there any movement on this? This would be an extremely useful feature for us...

**Comment by ccgagnon** [comment]
+1 It would be very useful feature for us too.

**Comment by galadriel2143** [comment]
@jantman @ccagnon Are either of you familiar with Terraform's test suite, and/or want to take a crack at it? Do you see anything that needs to be improved with my code? Please let me know.

**Comment by galadriel2143** [comment]
@ccgagnon

**Comment by galadriel2143** [comment]
Our team is very eager to have this merged in as well. Currently we're using the resources from a separate provider plugin.

**Comment by jantman** [comment]
Unfortunately I'm not familiar with Terraform's test suite, or even Go in general...

**Comment by galadriel2143** [comment]
@jantman I wasn't very familiar with Go either before I wrote this. Gogland is a pretty useful IDE if you need some assistance. I wouldn't recommend developing on Windows though.

**Comment by ccgagnon** [comment]
@galadriel2143 I am not familiar with Terraform's test suite, but I know go. I can take a look and try it.

**Comment by nikhil-p** [comment]
Any update on this? Need this feature soon.

**Comment by queeno** [comment]
I would like to get this feature in terraform asap as well

**Comment by bflad** [comment]
This has been released in [terraform-provider-aws version 1.9.0](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#190-february-09-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by lite2073** [comment]
Has anyone tried with an item that has nested attributes like a normal JSON would? I can't seem to make it work. Keep getting error: 

> Error: Invalid format of "item": Decoding failed: invalid character 'r' looking for beginning of value
